### PR TITLE
Support for specifying the Partner ID

### LINKED ISSUE #2393: Add user_agent argument to azurerm provider
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add user_agent attribute into the AzureRM provider.  If set then append value to the UserAgent header value.

The user_agent value in this context is used for adding a generated pid-<guid> to the telemetry. Having an argument in the provider would allow the value to be set within modules, or to set different values dependant on which provider alias is in effect.

### New or Affected Resource(s)

* [provider "AzureRM"](https://www.terraform.io/docs/providers/azurerm/index.html)

### Potential Terraform Configuration

```hcl
provider "azurerm" {
  user_agent = "pid-c8bdea86-f226-11e8-82db-bc838515ccd4"
}
```

### References

The AzureRM provider already appends the value of the AZURE_HTTP_USER_AGENT environment variable to the UserAgent header :

* [Append AZURE_HTTP_USER_AGENT environment variable to UserAgent string](https://github.com/terraform-providers/terraform-provider-azurerm/pull/483)
* [Override User-Agent for Azure partner customer attribution](https://github.com/terraform-providers/terraform-provider-azurerm/issues/2157)

Azure documentation on using GUIDs to attribute customer usage of solutions to the partner's offer and channel:
* <https://docs.microsoft.com/en-us/azure/marketplace/azure-partner-customer-usage-attribution#create-guids>

### PR DESCRIPTION
**Pull request by jeffreyCline**
(fixes #2393 )

### CODE REVIEW

File: azurerm/config.go [hunk 1]
```diff
@@ -370,12 +371,16 @@ func setUserAgent(client *autorest.Client) {
...
+		client.UserAgent = fmt.Sprintf("%s pid-%s", client.UserAgent, partnerID)
+	}
+
 	log.Printf("[DEBUG] AzureRM Client User Agent: %s\n", client.UserAgent)
 }
 
 // getArmClient is a helper method which returns a fully instantiated
 // *ArmClient based on the Config's current settings.
-func getArmClient(c *authentication.Config, skipProviderRegistration bool) (*ArmClient, error) {
+func getArmClient(c *authentication.Config, skipProviderRegistration bool, partnerId string) (*ArmClient, error) {
```
Comment (metacpp) [inline azurerm/config.go]: Since you always pass empty value to `partner_id` in test code, I feel you actually don't need to change getArmClient here.

**Comment by tombuildsstuff** [comment]
@jeffreyCline after chatting about this with @katbyte we're going to add a brief description into the "advanced options" section, which longer-term can point to a separate blog post for more info :)

File: azurerm/config.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/config.go]: getArmClient is used within the live provider and the test code - so this is necessary (whilst we could probably update the additional wrapper, we build clients in a couple of different manners depending on the requirements)

**Comment by tombuildsstuff** [comment]
It's green in Travis - no idea why it's still showing as pending ¯\_(ツ)_/¯

screenshot 2019-01-11 at 14 06 51

File: azurerm/config.go (hunk 1 above)
Comment (metacpp) [reply azurerm/config.go]: @tombuildsstuff you're correct, I missed the usage `provider.go` while reading it late night.

**Pull request merged**